File: platform-demos/C/guitar-tuner/guitar-tuner.py
#!/usr/bin/python

from gi.repository import Gtk, Gst, GObject
import os, sys

#Comment the first line and uncomment the second before installing
#or making the tarball (alternatively, use project variables)
UI_FILE = "src/guitar_tuner.ui"
#UI_FILE = "/usr/local/share/guitar_tuner_py/ui/guitar_tuner.ui"

class GUI:
	LENGTH = 500
	# Frequencies of the strings
	frequencies = {
		'E': 329.63,
		'A': 440,
		'D': 587.33,
		'G': 783.99,
		'B': 987.77,
		'e': 1318.5
	}

	def __init__(self):
		self.builder = Gtk.Builder()
		self.builder.add_from_file(UI_FILE)
		self.builder.connect_signals(self)

		window = self.builder.get_object('window')
		window.show_all()

	def on_button_clicked (self, button):
		label = button.get_child()
		text = label.get_label()
		self.play_sound (self.frequencies[text])

	def destroy(window, self):
		Gtk.main_quit()

	def pipeline_stop(self, pipeline):
		pipeline.set_state(Gst.State.NULL)
		return False

	def play_sound(self, frequency):
		pipeline = Gst.Pipeline(name='note')
		source = Gst.ElementFactory.make('audiotestsrc', 'src')
		sink = Gst.ElementFactory.make('autoaudiosink', 'output')

		source.set_property('freq', frequency)
		pipeline.add(source)
		pipeline.add(sink)
		source.link(sink)

		pipeline.set_state(Gst.State.PLAYING)

		GObject.timeout_add(self.LENGTH, self.pipeline_stop, pipeline)

def main():
	Gst.init_check(sys.argv)
	app = GUI()
	Gtk.main()

if __name__ == "__main__":
    sys.exit(main())

File: platform-demos/C/guitar-tuner/guitar-tuner.c
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 4; tab-width: 4 -*- */

#include <config.h>
#include <gtk/gtk.h>
#include <gst/gst.h>
#include <glib/gi18n.h>

/* For testing purposes, use the local (not installed) ui file */
/* #define UI_FILE PACKAGE_DATA_DIR"/guitar_tuner/ui/guitar_tuner.ui" */
#define UI_FILE "src/guitar_tuner.ui"

/* Frequencies of the tones */
#define NOTE_E 329.63
#define NOTE_A 440
#define NOTE_D 587.33
#define NOTE_G 783.99
#define NOTE_B 987.77
#define NOTE_e 1318.5

#define LENGTH 500

#define WIDGET_WINDOW "window"

static gboolean
pipeline_stop (GstElement* pipeline)
{
	gst_element_set_state (pipeline, GST_STATE_NULL);
	g_object_unref (pipeline);

	/* disconnect handler */
	return FALSE;
}

static void 
play_sound (gdouble frequency)
{
	GstElement *source, *sink;
	GstElement *pipeline;

	pipeline = gst_pipeline_new ("note");
	source   = gst_element_factory_make ("audiotestsrc",
	                                     "source");
	sink     = gst_element_factory_make ("autoaudiosink",
	                                     "output");

	/* set frequency */
	g_object_set (source, "freq", frequency, NULL);
	
	gst_bin_add_many (GST_BIN (pipeline), source, sink, NULL);
	gst_element_link (source, sink);

	gst_element_set_state (pipeline, GST_STATE_PLAYING);

	/* stop it after 500ms */
	g_timeout_add (LENGTH, (GSourceFunc) pipeline_stop, pipeline);
}

/* Callback for the buttons */
void on_button_clicked (GtkButton* button,
                        gpointer user_data)
{
	const gchar* text = gtk_button_get_label (button);

	if (g_str_equal (text, _("E")))
	    play_sound (NOTE_E);
	else if (g_str_equal (text, _("A")))
	    play_sound (NOTE_A);
	else if (g_str_equal (text, _("G")))
	    play_sound (NOTE_G);
	else if (g_str_equal (text, _("D")))
	    play_sound (NOTE_D);
	else if (g_str_equal (text, _("B")))
	    play_sound (NOTE_B);
	else if (g_str_equal (text, _("e")))
	    play_sound (NOTE_e);
}

int
main (int argc, char *argv[])
{
 	GtkWidget *window;
	GtkBuilder *builder;

	/* Initialize gtk+ and gstreamer */
	gst_init (&argc, &argv);
	gtk_init (&argc, &argv);

	builder = gtk_builder_new ();
	gtk_builder_add_from_file (builder, UI_FILE, NULL);

	window = GTK_WIDGET (gtk_builder_get_object (builder, WIDGET_WINDOW));
	gtk_builder_connect_signals (builder, NULL);
	
	gtk_widget_show (window);

	gtk_main ();
	return 0;
}
